libs/blocks/preflight/panels/assets.js

```diff
@@ -0,0 +1,188 @@
+import { html, signal, useEffect } from '../../../deps/htm-preact.js';
+import { createTag } from '../../../utils/utils.js';
+
+function isViewportTooSmall() {
+  return !window.matchMedia('(min-width: 1200px)').matches;
+}
+
+const imagesWithMismatch = signal([]);
+const imagesWithMatch = signal([]);
+const checksPerformed = signal(false);
+const viewportTooSmall = signal(isViewportTooSmall());
+
+const groups = [
+  { title: 'Images with dimension mismatch', imgArray: imagesWithMismatch },
+  { title: 'Images with matching dimensions', imgArray: imagesWithMatch },
+];
+
+async function checkImageDimensions() {
+  if (checksPerformed.value || viewportTooSmall.value) return;
+  checksPerformed.value = true;
+
+  const allImages = [...document.querySelectorAll('main picture img')];
+  if (!allImages.length) return;
+
+  // Force load all images
+  await Promise.all(
+    allImages.map((img) => {
+      if (!img.complete) {
+        img.setAttribute('loading', 'eager');
+        return new Promise((resolve) => {
+          img.addEventListener('load', resolve);
+        });
+      }
+      return Promise.resolve();
+    }),
+  );
+
+  // Filter the loaded images to ensure they are visible, not an icon, and not an SVG
+  const images = allImages.filter((img) => img.checkVisibility()
+    && !img.closest('.icon-area')
+    && !img.src.endsWith('.svg'));
+
+  if (!images.length) return;
+
+  const maxFullWidth = 1920;
+  const viewportWidth = document.documentElement.clientWidth;
+
+  for (const img of images) {
+    // Get the original dimensions of the uploaded image; fallback to the natural dimensions
+    const naturalWidth = img.getAttribute('width') ? parseInt(img.getAttribute('width'), 10) : img.naturalWidth;
+    const naturalHeight = img.getAttribute('height') ? parseInt(img.getAttribute('height'), 10) : img.naturalHeight;
+    // Get the display dimensions of the image
+    const displayWidth = img.offsetWidth;
```

Comment:
Not a big deal, but this would be more efficient if you just got the getBoundingClientRect() to avoid layout thrashing. 
Getting the offsetWidth and offsetHeight separately causes two different DOM calculations. 

```
const rect = img.getBoundingClientRect();
const displayWidth = rect.width;
const displayHeight = rect.height;
```

Reply:
This is quite debatable, we've had requests in the past to move away from that approach, as it was leading to noticeable reflow issues. This is the first article I could find, but there are a bunch more - https://toruskit.com/blog/how-to-get-element-bounds-without-reflow. One could argue that we can move to using an IntersectionObserver, but I think that's overkill for this particular context of non-customer facing logic.

Reply:
That's cool, I didn't consider an intersection observer as well. 
yes, this is definitely a nit. :)